tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -554,6 +554,93 @@ int main(int argc, char **argv)
+     * been received if TLS 1.3 was negotiated.
+     *
+     * This behavior differs between TLS 1.2 and TLS 1.3, so negotiate TLS 1.3 with
+     * a TLS 1.2 session ticket and assert that the TLS 1.3 connection still
+     * behaves correctly.
+     */
+    {
+        DEFER_CLEANUP(struct s2n_blob ticket = { 0 }, s2n_free);
+
+        struct s2n_connection *client_conn = s2n_connection_new(S2N_CLIENT);
+        struct s2n_connection *server_conn = s2n_connection_new(S2N_SERVER);
+        EXPECT_NOT_NULL(client_conn);
+        EXPECT_NOT_NULL(server_conn);
+
+        EXPECT_SUCCESS(s2n_connection_set_config(client_conn, tls13_client_config));
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, tls12_server_config));
+
+        /* Create nonblocking pipes */
+        struct s2n_test_io_pair io_pair = { 0 };
+        EXPECT_SUCCESS(s2n_io_pair_init_non_blocking(&io_pair));
+        EXPECT_SUCCESS(s2n_connections_set_io_pair(client_conn, server_conn, &io_pair));
+
+        /* Negotiate initial handshake to produce TLS1.2 session ticket */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_TRUE(IS_ISSUING_NEW_SESSION_TICKET(server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS12);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS12);
+
+        /* Store the TLS1.2 session ticket */
+        size_t tls12_session_ticket_len = s2n_connection_get_session_length(client_conn);
+        EXPECT_SUCCESS(s2n_alloc(&ticket, tls12_session_ticket_len));
+        EXPECT_SUCCESS(s2n_connection_get_session(client_conn, ticket.data, ticket.size));
+
+        /* Prepare client and server for a second connection */
+        EXPECT_OK(s2n_wipe_connections(client_conn, server_conn, &io_pair));
+
+        /* Client sets up a resumption connection with the received TLS1.2 session ticket data */
+        EXPECT_SUCCESS(s2n_connection_set_session(client_conn, ticket.data, ticket.size));
+
+        /* Set server config to TLS1.3 cipher preferences to allow TLS 1.3 */
+        EXPECT_SUCCESS(s2n_connection_set_config(server_conn, server_config));
+
+        /* Negotiate second connection */
+        EXPECT_SUCCESS(s2n_negotiate_test_server_and_client(server_conn, client_conn));
+        EXPECT_TRUE(ARE_FULL_HANDSHAKES(client_conn, server_conn));
+        EXPECT_EQUAL(server_conn->actual_protocol_version, S2N_TLS13);
+        EXPECT_EQUAL(client_conn->actual_protocol_version, S2N_TLS13);
+
+        /* since TLS 1.3 was negotiated, and the client hasn't called recv yet
+         * there should be no session ticket available. 512 is an arbitrary
+         * number, nothing should be copied because this operation should fail
+         */
+        EXPECT_EQUAL(s2n_connection_get_session_length(client_conn), 0);
+        EXPECT_EQUAL(s2n_connection_get_session(client_conn, ticket.data, 512), 0);
```

Comment:
If you're not using ticket.size, you need to ensure that 512<=ticket.size. Otherwise this is very dangerous.

Reply:
My reason for writing the asserts this way is that clients might not be calling `get_session_length` before calling `get_session`, so we should independently assert that `get_session` also fails. 

I'm not sure I quite understand that danger in the current test. Are you saying that in general this is a dangerous pattern (definitely agreed) or that it is dangerous to be doing this for this particular test? I think it is safe to do this in this test, because we assert that no data is actually getting written. If any data is getting written, this test will fail.

And for the general case, shouldn't we assert that `512>=ticket.size`?

Reply:
It's about memory safety. In C, pointers do not have a size associated with them. s2n_connection_get_session takes memory and the size of the memory. By telling s2n that ticket.data has a size of 512 without checking that that's true, you could wrong and ticket.data could be <512, which won't lead to an error, it'll lead to a segfault. This is like doing s2n_connection_get_session(client_conn, NULL, 512) because you don't expect us to actually write anything.

If you're wrong, you want an error, not a segfault.

Reply:
Basically: NEVER lie about the size of memory in C. Tests can lie about a lot of things, but not that :)